cnd/src/http_api/routes.rs

```diff
@@ -154,137 +181,99 @@ where
-
-                    if refund.is_some() {
-                        tracing::debug!("We have decided it's time to refund.");
-                        refund
-                    } else {
-                        let action = possible_next_actions.iter()
-                            .filter(|action_name| **action_name != ActionName::Refund)
-                            .collect::<Vec<_>>()
-                            .get(0)
-                            .copied()
-                            .copied();
-
-                        tracing::debug!("Remaining action is {:?}", action);
-                        action
-                    }
-                }
-                _ => unreachable!("Unexpected error encountered, there should not be more than 2 actions available at a given time"),
-            };
+    Ok(())
+}
 
-            tracing::debug!("Next action is {:?}", next_action);
+fn add_events(
+    entity: &mut siren::Entity,
+    alpha_events: LedgerEvents,
+    beta_events: LedgerEvents,
+) -> anyhow::Result<()> {
+    let alpha_state_sub = siren::SubEntity::from_entity(
+        siren::Entity::default()
+            .with_class_member("state")
+            .with_properties(alpha_events)
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
+        &["alpha"],
+    );
+    entity.push_sub_entity(alpha_state_sub);
 
-            if let Some(action_name) = next_action {
-                let siren_action = make_siren_action(swap_id, action_name);
-                return Ok(entity.with_action(siren_action));
-            }
+    let beta_state_sub = siren::SubEntity::from_entity(
+        siren::Entity::default()
+            .with_class_member("state")
+            .with_properties(beta_events)
+            .map_err(|e| {
+                tracing::error!("failed to set properties of entity: {:?}", e);
+                HttpApiProblem::with_title_and_type_from_status(StatusCode::INTERNAL_SERVER_ERROR)
+            })?,
+        &["beta"],
+    );
+    entity.push_sub_entity(beta_state_sub);
 
-            Ok(entity)
-        }
-        _ => Ok(entity),
-    }
+    Ok(())
 }
 
-fn filter_for_next_possible_actions(available: Vec<ActionName>) -> Vec<ActionName> {
-    tracing::debug!("Available actions: {:?}", available);
-    if available.contains(&ActionName::Init) {
-        return vec![ActionName::Init];
+async fn next_available_action<S>(facade: Facade, swap: &S) -> anyhow::Result<Option<ActionName>>
+where
+    S: GetRole
+        + DeployAction
+        + InitAction
+        + FundAction
+        + RedeemAction
+        + RefundAction
+        + Clone
+        + AlphaBlockchain
+        + BetaBlockchain
+        + AlphaAbsoluteExpiry
+        + BetaAbsoluteExpiry,
+{
+    if swap.init_action().is_ok() {
+        return Ok(Some(ActionName::Init));
     }
```

Comment:
I was about to comment in the other patch that this logic is pretty confusing and thankfully, you are refactoring it here :D

Would have preferred to have this in one patch if patch-by-patch review is being recommended :)
As you once said @tcharding, "I don't care about the steps it took you to come to the solution" :)

Reply:
Yeah, I was lazy. The initial refactor of this function _should_ have been done as the initial patch and then only the additional logic would be in the final patch. But by the time I did the refactor I was too lazy to rebase and do it properly.